### PR TITLE
[ MO ] Support MXNet operation arange_like

### PR DESCRIPTION
**Pull request by yekruglov**
Description:
Support MXNet [arange_like](https://mxnet.apache.org/versions/1.6/api/python/docs/api/ndarray/contrib/index.html#mxnet.ndarray.contrib.arange_like) operation by the Model Optimizer tool

Ticket: 61899

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests
* [x]  Transformation tests
* [x]  Model Optimizer IR Reader check N/A

Documentation:
* [x]  Supported frameworks operations list N/A
* [x]  Guide on how to convert the **public** model N/A
* [x]  User guide update N/A

### CODE REVIEW

File: model-optimizer/extensions/front/mxnet/arange_like_replacer.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        # we have to correct the stop value for the Range node if start > 0
+        if node.start > 0:
+            correct_stop_value = create_op_with_const_inputs(graph, Add, port_value_dict={1: int64_array(node.start)},
+                                                             op_attrs={'name': range_node.name + '/Stop'})
+            range_node.in_port(1).get_connection().insert_node(correct_stop_value)
+
+        # Range node supports only scalar inputs
+        squeeze_node = create_op_with_const_inputs(graph, Squeeze, port_value_dict={1: int64_array(0)},
+                                                   op_attrs={"name": range_node.name + '/Stop/Squeeze'})
+        range_node.in_port(1).get_connection().insert_node(squeeze_node)
```
Comment (sadolini) [inline model-optimizer/extensions/front/mxnet/arange_like_replacer.py:80]: Add renaming last node in subgraph to name of operation node to save framework name of operation after transformation.

File: model-optimizer/extensions/front/mxnet/arange_like_replacer.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+from mo.ops.squeeze import Squeeze
+from mo.utils.error import Error
+
+
+class ArangeLikeReplacer(FrontReplacementOp):
+    op = 'arange_like'
+    enabled = True
+
+    def replace_sub_graph(self, graph: Graph, match: dict):
+        node = match['op']
```
Comment (sadolini) [inline model-optimizer/extensions/front/mxnet/arange_like_replacer.py:24]: You dont't use step attribute of operation. Do we support only default value of attribute? Then we should check it somewhere.

File: model-optimizer/extensions/front/mxnet/arange_like_replacer.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        if node.repeat != 1:
+            raise Error("arange_like node {} with non default repeat value {} "
+                        "is not supported".format(name, node.repeat))
+        axis = node.axis
+        shape_node = Shape(graph, {'name': name + '/shape'}).create_node()
+        range_node = create_op_with_const_inputs(graph, Range, {0: int64_array(node.start),
+                                                                2: int64_array(1)}, {'name': name + '/Range'})
+        node.in_port(0).get_connection().set_destination(shape_node.in_port(0))
+
+        if axis:
```
Comment (sadolini) [inline model-optimizer/extensions/front/mxnet/arange_like_replacer.py:35]: if axis is 0, which branch should work? In tests you check axis None and not None (3), should we check here that axis is not None?

File: model-optimizer/extensions/front/mxnet/arange_like_replacer.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+from mo.front.common.replacement import FrontReplacementOp
+from mo.front.tf.graph_utils import create_op_with_const_inputs
+from mo.graph.graph import Graph
+from mo.ops.shape import Shape
+from mo.ops.reshape import Reshape
+from mo.ops.squeeze import Squeeze
+from mo.utils.error import Error
+
+
+class ArangeLikeReplacer(FrontReplacementOp):
```
Comment (sadolini) [inline model-optimizer/extensions/front/mxnet/arange_like_replacer.py:19]: Please, add description what operation do

**Review by sadolini — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/front/mxnet/arange_like_replacer.py (hunk 1 above)
Comment (yekruglov) [reply model-optimizer/extensions/front/mxnet/arange_like_replacer.py]: Done

File: model-optimizer/extensions/front/mxnet/arange_like_replacer.py (hunk 2 above)
Comment (yekruglov) [reply model-optimizer/extensions/front/mxnet/arange_like_replacer.py]: Missed it. Done

File: model-optimizer/extensions/front/mxnet/arange_like_replacer.py (hunk 3 above)
Comment (yekruglov) [reply model-optimizer/extensions/front/mxnet/arange_like_replacer.py]: Done

File: model-optimizer/extensions/front/mxnet/arange_like_replacer.py (hunk 4 above)
Comment (yekruglov) [reply model-optimizer/extensions/front/mxnet/arange_like_replacer.py]: Added description to mo/ops/arange_like.py

File: tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            correct_stop_value = create_op_with_const_inputs(graph, Add, port_value_dict={1: int64_array(node.start)},
+                                                             op_attrs={'name': range_node.name + '/Stop'})
+            range_node.in_port(1).get_connection().insert_node(correct_stop_value)
+
+        # Range node supports only scalar inputs
+        squeeze_node = create_op_with_const_inputs(graph, Squeeze, port_value_dict={1: int64_array(0)},
+                                                   op_attrs={"name": range_node.name + '/Stop/Squeeze'})
+        range_node.in_port(1).get_connection().insert_node(squeeze_node)
+
+        rename_nodes([(node, name + '/ShouldBeDeleted'), (range_node, name)])
```
Comment (sadolini) [inline tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py:79]: range node is not last for case axis = None, reshape should be renamed for this case

File: tools/mo/openvino/tools/mo/front/mxnet/arange_like_ext.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# Copyright (C) 2018-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import numpy as np
+
+from openvino.tools.mo.front.extractor import FrontExtractorOp
+from openvino.tools.mo.front.mxnet.extractors.utils import get_mxnet_layer_attrs
+from openvino.tools.mo.ops.arange_like import ArangeLikeOp
```
Comment (sadolini) [inline tools/mo/openvino/tools/mo/front/mxnet/arange_like_ext.py:8]: sort imports

Comment (yekruglov) [reply tools/mo/openvino/tools/mo/front/mxnet/arange_like_ext.py:8]: It is sorted, isn't it?

Comment (sadolini) [reply tools/mo/openvino/tools/mo/front/mxnet/arange_like_ext.py:8]: lines 8 and 9 )

File: tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+from openvino.tools.mo.ops.shape import Shape
+from openvino.tools.mo.ops.squeeze import Squeeze
+from openvino.tools.mo.utils.error import Error
+
+
+class ArangeLikeReplacer(FrontReplacementOp):
+    op = 'arange_like'
+    enabled = True
+
+    def replace_sub_graph(self, graph: Graph, match: dict):
```
Comment (sadolini) [inline tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py:20]: nit: add coment what sub-graph will be created

File: tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            shape_node.out_port(0).connect(reshape_node1.in_port(0))
+            reshape_node1.out_port(0).connect(range_node.in_port(1))
+            range_node.out_port(0).connect(reshape_node2.in_port(0))
+            shape_node.out_port(0).connect(reshape_node2.in_port(1))
+            node.out_port(0).get_connection().set_source(reshape_node2.out_port(0))
+            rename_nodes([(node, name + '/ShouldBeDeleted'), (reshape_node2, name)])
+
+        # MXNet arange_like op has no stop attribute and the result tensor always matches the input shape, so
+        # we have to correct the stop value for the Range node if start > 0
+        if node.start > 0:
```
Comment (sadolini) [inline tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py:71]: minor: start can't be negative, right? Should we check it somewhere (assert)?

File: tools/mo/unit_tests/mo/front/mxnet/arange_like_replacer_test.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+from openvino.tools.mo.front.common.partial_infer.utils import int64_array
+from openvino.tools.mo.front.mxnet.arange_like_replacer import ArangeLikeReplacer
+from openvino.tools.mo.utils.error import Error
+from openvino.tools.mo.utils.ir_engine.compare_graphs import compare_graphs
+from unit_tests.utils.graph import build_graph, shaped_parameter, regular_op_with_empty_data, result, connect, \
+    shaped_const_with_data, connect_data
+
+
+class ArangeLikeReplacerTest(unittest.TestCase):
```
Comment (sadolini) [inline tools/mo/unit_tests/mo/front/mxnet/arange_like_replacer_test.py:14]: Could we have axis<0? If yes, please add such test case.

File: tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py (hunk 8 above)
Comment (yekruglov) [reply tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py:71]: I guess, there should be `if node.start != 0`

File: tools/mo/openvino/tools/mo/front/mxnet/arange_like_ext.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+    @classmethod
+    def extract(cls, node: Node):
+        attrs = get_mxnet_layer_attrs(node.symbol_dict)
+        ArangeLikeOp.update_node_stat(node, {
+            'start': attrs.float('start', 0),
+            'repeat': attrs.int('repeat', 1),
+            'step': attrs.float('step', 1),
+            'axis': attrs.int('axis', None),
+            'dtype': np.dtype(attrs.str('dtype ', 'float32'))
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/front/mxnet/arange_like_ext.py:24]: why is this attribute needed?

File: tools/mo/openvino/tools/mo/ops/gather.py [hunk 11]
```diff
@@ -89,6 +89,7 @@ def infer(node: Node):
         data_value = node.in_port(0).data.get_value()
         indices_value = node.in_port(1).data.get_value()
         if data_value is not None and indices_value is not None and is_fully_defined(indices_value):
+            indices_value = int64_array(indices_value)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/ops/gather.py:92]: why is this change needed if we care of a type of indices overall?

File: tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+            reshape_forward_node = create_op_with_const_inputs(graph, Reshape, {1: int64_array([-1])},
+                                                               {'name': name + '/Reshape_forward'})
+
+            reshape_backward_node = Reshape(graph, {'name': name + '/Reshape_backward'}).create_node()
+            shape_of_node = Shape(graph, {'name': reshape_forward_node.name + '/ShapeOf'}).create_node()
+
+            reshape_forward_node.in_port(0).connect(data_out_port)
+            reshape_forward_node.out_port(0).connect(shape_of_node.in_port(0))
+            shape_of_node.out_port(0).connect(range_node.in_port(1))
```
Comment (pavel-esir) [inline tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py:71]: nit: you insert below Squeeze for `in_port(0)` but for `in_port(1)` you don't do that. For me it's unclear why this works since `ShapeOf` according to spec always returnd 1D array not a scalar

File: tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+            Sub-graph after Range node will be look like
+            
+            Range - Reshape([-1, 1]) - Tile([1, repeat]) - Reshape(-1) - Slice
+            
+            """
+            div_node = create_op_with_const_inputs(graph, Div, {1: int64_array([node.repeat])},
+                                                   {'name': name + '/Divide'})
+            add_node = create_op_with_const_inputs(graph, Add, {1: int64_array([1])},
+                                                   {'name': div_node.name + '/Add'})
+            cast_node = Cast(graph, {'name': name + '/ConvertToI64', 'dst_type': np.int64}).create_node()
```
Comment (pavel-esir) [inline tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py:92]: nit: this is not necessary but if you have enough time please add this node into subgraph picture in line 50

**Review by pavel-esir — CHANGES_REQUESTED** [review]
In general looks good for me. Left some comments. Applying nit comments is optional not required

File: tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py (hunk 12 above)
Comment (pavel-esir) [reply tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py:71]: Discussed personally. `in_port(0)` of Range is always a scalar constant regardless of `node.start` value. Comment resolved.

File: tools/mo/openvino/tools/mo/front/mxnet/arange_like_ext.py (hunk 10 above)
Comment (yekruglov) [reply tools/mo/openvino/tools/mo/front/mxnet/arange_like_ext.py]: Removed

File: tools/mo/openvino/tools/mo/ops/gather.py (hunk 11 above)
Comment (yekruglov) [reply tools/mo/openvino/tools/mo/ops/gather.py:92]: np.ma.take below raise an exception if `indicies_value` has another type. Found this case in the customer model, there was an array with `dtype` = `float`.

File: tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py (hunk 13 above)
Comment (yekruglov) [reply tools/mo/openvino/tools/mo/front/mxnet/arange_like_replacer.py:92]: This node does not belong to the subgraph in line 50

**Review by andrew-zaytsev — APPROVED** [review]

**Review by pavel-esir — APPROVED** [review]

**Review by evolosen — APPROVED** [review]

**Review by sadolini — APPROVED** [review]

**Comment by rkazants** [comment]
@yekruglov, please check why one MXNet e2e test is failing:
```
<html>
<body>
<!--StartFragment-->

Failed | test.py::test_run[MXNET_DeformableConvRfcn_api_2_False_batch_1_device_CPU_fusing_True_precision_FP32] | 306.27 |  
-- | -- | -- | --
[gw3] linux -- Python 3.6.15 /home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/tmp/e2e_test_venv/bin/pythonTraceback (most recent call last):  File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/repos/frameworks.ai.openvino.tests/utils/multiprocessing_utils.py", line 58, in multiprocessing_run    error_message, *ret_args = queue.get(timeout=timeout)  File "/usr/lib/python3.6/multiprocessing/queues.py", line 105, in get    raise Emptyqueue.EmptyDuring handling of the above exception, another exception occurred:Traceback (most recent call last):  File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/repos/frameworks.ai.openvino.tests/utils/multiprocessing_utils.py", line 60, in multiprocessing_run    raise TimeoutError("{func} running timed out!".format(func=func_log_name))multiprocessing.context.TimeoutError: Inference Engine Python API running timed out!During handling of the above exception, another exception occurred:Traceback (most recent call last):  File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/repos/frameworks.ai.openvino.tests/e2e_oss/test.py", line 138, in test_run    ie_pipeline.run()  File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/repos/frameworks.ai.openvino.tests/utils/e2e/common/pipeline.py", line 24, in run    step.execute(step_input)  File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/repos/frameworks.ai.openvino.tests/utils/e2e/infer/provider.py", line 28, in execute    self.out_data = self.executor.infer(data)  File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/repos/frameworks.ai.openvino.tests/utils/e2e/infer/common_inference.py", line 145, in infer    multiprocessing_run(self._infer, [input_data], "Inference Engine Python API", self.timeout)  File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/repos/frameworks.ai.openvino.tests/utils/multiprocessing_utils.py", line 77, in multiprocessing_run    "{func} was killed with a signal {signal}".format(func=func_log_name, signal=exit_signal))multiprocessing.context.ProcessError: Inference Engine Python API was killed with a signal Signals.SIGSEGV----------------------------- Captured stdout call -----------------------------[ INFO ] Running MXNET_DeformableConvRfcn_api_2_False_batch_1_device_CPU_fusing_True_precision_FP32 test [ INFO ] Running reference pipeline: [ INFO ] Reading references from path /home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/repos/frameworks.ai.openvino.tests/e2e_oss/test_data/references/mxnet/def_conv.npz [ INFO ] Running inference pipeline: [ INFO ] Reading input file from /home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/repos/frameworks.ai.openvino.tests/e2e_oss/test_data/inputs/mxnet/data_def_conv.npz ... [ INFO ] Align batch data for layers "data", "data2" to batch 1 ... [ INFO ] Running Model Optimizer: /home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/tmp/e2e_test_venv/bin/mo --input_model /nfs/inn/proj/vdp/vdp_tests/models/internal/mxnet/deformable_rfcn/def_conv-0000.params --output_dir /home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux- ubuntu20-e2e/b/install/dldt_cpack/out_dir/MXNET_DeformableConvRfcn_api_2_False_batch_1_device_CPU_fusing_True_precision_FP32h3qo_5dl --data_type FP32 --input data,data2 --input_shape [1,512,63,38],[1,512,63,38] [ INFO ] Model Optimizer arguments: Common parameters: 	- Path to the Input Model: 	/nfs/inn/proj/vdp/vdp_tests/models/internal/mxnet/deformable_rfcn/def_conv-0000.params 	- Path for generated IR: 	/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/install/dldt_cpack/out_dir/MXNET_DeformableConvRfcn_api_2_False_batch_1_device_CPU_fusing_True_precision_FP32h3qo_5dl 	- IR output name: 	def_conv-0000 	- Log level: 	ERROR 	- Batch: 	Not specified, inherited from the model 	- Input layers: 	data,data2 	- Output layers: 	Not specified, inherited from the model 	- Input shapes: 	[1,512,63,38],[1,512,63,38] 	- Source layout: 	Not specified 	- Target layout: 	Not specified 	- Layout: 	Not specified 	- Mean values: 	Not specified 	- Scale values: 	Not specified 	- Scale factor: 	Not specified 	- Precision of IR: 	FP32 	- Enable fusing: 	True 	- Enable grouped convolutions fusing: 	True 	- Move mean values to preprocess section: 	None 	- Reverse input channels: 	False 	- Use legacy API for model processing: 	False 	- Use the transformations config file: 	None MXNet specific parameters: 	- Deploy-ready symbol file: 	None 	- Enable MXNet loader for models trained with MXNet version lower than 1.0.0: 	False 	- Prefix name for args.nd and argx.nd files: 	None 	- Pretrained model to be merged with the .nd files: 	None 	- Enable saving built parameters file from .nd files: 	False 	- OpenVINO runtime found in: 	/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/tmp/e2e_test_venv/lib/python3.6/site-packages/openvino OpenVINO runtime version: 	2022.1.0-5547-c04d85e0268-refs/pull/8939/head Model Optimizer version: 	2022.1.0-5547-c04d85e0268-refs/pull/8939/head [ SUCCESS ] Generated IR version 11 model. [ SUCCESS ] XML file: /home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/install/dldt_cpack/out_dir/MXNET_DeformableConvRfcn_api_2_False_batch_1_device_CPU_fusing_True_precision_FP32h3qo_5dl/def_conv-0000.xml [ SUCCESS ] BIN file: /home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/install/dldt_cpack/out_dir/MXNET_DeformableConvRfcn_api_2_False_batch_1_device_CPU_fusing_True_precision_FP32h3qo_5dl/def_conv-0000.bin [ SUCCESS ] Total execution time: 3.92 seconds.  [ SUCCESS ] Memory consumed: 240 MB.   [ ERROR ] [ WARNING ]   Detected not satisfied dependencies: 	mxnet: installed: 1.7.0, required: ~= 1.7.0.post2 	networkx: installed: 2.5.1, required: ~= 2.6  Please install required versions of components or use install_prerequisites script /home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/tmp/e2e_test_venv/lib/python3.6/site-packages/openvino/tools/mo/install_prerequisites/install_prerequisites_mxnet.sh Note that install_prerequisites scripts may install additional components.  [ INFO ] Using old API [ INFO ] Inference Engine version: 2022.1.0-5547-c04d85e0268-refs/pull/8939/head [ INFO ] Creating IE Core Engine... [ INFO ] Loading network files [ INFO ] IE Network has the same batch as requested to set. Batch size setting skipped. [ INFO ] Loading network to the CPU device... [ INFO ] Starting inference [ INFO ] Inference run counter: 0 ----------------------------- Captured stderr call -----------------------------*** Segmentation fault Register dump:   RAX: 0000000000000000   RBX: 0000000000000080   RCX: 00007f4e7908724b  RDX: 0000000000000000   RSI: 0000000000e68d60   RDI: 0000000000000002  RBP: 00007f4d7c2182c0   R8 : 0000000000000000   R9 : 0000000000e68d60  R10: 0000000000000008   R11: 0000000000000246   R12: 000000000000000

<!--EndFragment-->
</body>
</html>
```

**Comment by pavel-esir** [comment]
> @yekruglov, please check why one MXNet e2e test is failing with seg fault:
> 
> ```
> test.py::test_run[MXNET_DeformableConvRfcn_api_2_False_batch_1_device_CPU_fusing_True_precision_FP32]
> ```

I had the same failure with this mxnet model. Seems like it's a sporadic issue since after Rebuild without any changes the test passed

**Review by rkazants — APPROVED** [review]
